File: scripts/optimize-images.js
const path = require('path');
const fs = require('fs');
// glob will be imported using modern API later

let sharp;
try {
  sharp = require('sharp');
} catch (e) {
  console.error('The "sharp" module is not installed. Install it with "npm install --save-dev sharp" to enable image optimization. Exiting.');
  process.exit(0);
}

const inputDir = path.normalize(path.join(__dirname, '..', 'assets'));
const outputDir = path.join(inputDir, 'optimized');

if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

const sizes = [480, 800, 1200];

async function processFile(file) {
  const ext = path.extname(file).toLowerCase();
  const base = path.basename(file, ext);
  const img = sharp(file);
  const metadata = await img.metadata();

  // create webp (main)
  const webpPath = path.join(outputDir, `${base}.webp`);
  await img.webp({ quality: 80 }).toFile(webpPath);

  // create sizes where original is larger than target
  await Promise.all(sizes.map(async (w) => {
    if (metadata.width && metadata.width <= w) return; // skip if original is smaller
    const outPath = path.join(outputDir, `${base}-${w}.webp`);
    await img.resize({ width: w }).webp({ quality: 80 }).toFile(outPath);
  }));

  // also copy original file into optimized folder with suffix -orig
  const fallbackPath = path.join(outputDir, `${base}-orig${ext}`);
  await img.toFile(fallbackPath);

  console.log('Processed', file);
}

const { glob } = require('glob');

async function main() {
  try {
    const pattern = path.join(inputDir, '*.+(png|jpg|jpeg)');
    const files = await glob(pattern, { nocase: true, nodir: true });
    if (!files || files.length === 0) {
      console.log('No image files found in', inputDir);
      return;
    }
    for (const f of files) {
      try {
        await processFile(f);
      } catch (e) {
        console.error('Error processing', f, e.message);
      }
    }
    console.log('Done optimizing images. Output ->', outputDir);
  } catch (err) {
    console.error('Error while searching for images:', err.message || err);
  }
}

main();

File: scripts/optimize-images.py
from PIL import Image
import os

SIZES = [480, 800, 1200]

BASE_DIR = os.path.join(os.path.dirname(__file__), '..')
ASSETS_DIR = os.path.join(BASE_DIR, 'assets')
OUT_DIR = os.path.join(ASSETS_DIR, 'optimized')

os.makedirs(OUT_DIR, exist_ok=True)

def process_image(path):
    base, ext = os.path.splitext(os.path.basename(path))
    try:
        with Image.open(path) as im:
            im_format = im.format
            # Save main webp
            webp_path = os.path.join(OUT_DIR, f"{base}.webp")
            im.save(webp_path, 'WEBP', quality=80)
            print('Saved', webp_path)

            # Create resized webp versions
            for w in SIZES:
                if im.width <= w:
                    continue
                im_copy = im.copy()
                im_copy.thumbnail((w, w), Image.LANCZOS)
                out_path = os.path.join(OUT_DIR, f"{base}-{w}.webp")
                im_copy.save(out_path, 'WEBP', quality=80)
                print('Saved', out_path)

            # Save a fallback optimized copy of original type
            fallback = os.path.join(OUT_DIR, f"{base}-orig{ext}")
            im.save(fallback)
            print('Saved', fallback)
    except Exception as e:
        print('Error processing', path, e)

if __name__ == '__main__':
    for fname in os.listdir(ASSETS_DIR):
        if fname.lower().endswith(('.png', '.jpg', '.jpeg')):
            process_image(os.path.join(ASSETS_DIR, fname))

    print('Done. Optimized images are in', OUT_DIR)
